satpy/scene.py

```diff
@@ -803,11 +803,31 @@ def __getitem__(self, key):
 
     def __setitem__(self, key, value):
         """Add the item to the scene."""
+        if key in self:
+            old_key = self._datasets.get_key(key)
+            self._wishlist.discard(old_key)
+            del self._datasets[old_key]
+
+        if isinstance(value, np.ndarray):
+            value = xr.DataArray(value)
+
+        name = key
+        if isinstance(key, DataID):
+            name = key['name']
+
+        if value.attrs.get('name', None) is None or value.attrs['name'] != name:
+            value.attrs['name'] = name
+
+        value.attrs.pop('_satpy_id', None)
```

Comment:
Should this not be popped so that `DatasetDict.__setitem__` can use the id keys config from it?

Reply:
I hope I remember this correctly; At first I tried that but then found out that the `DataID.from_dataarray` class method always returned the `_satpy_id` directly if present so it wouldn't get updated.

Reply:
:thinking: But after the `self._datasets[key] = value` shouldn't the new `_satpy_id` be set?

Reply:
Yes but I call `DataID.from_dataarray` in the `datasets.__setitem__` method and therefore I need to remove it because otherwise `DataID.from_dataarray` will return the old id. This could probably be done different though?

Reply:
Good point, but then I think this pop should go inside `DatasetDict.__setitem__`, right?

And actually, instead of `from_dataarray` you could use `data_id.from_dict` if you have the original DataID of the DataArray...but yeah if the provided DataArray didn't already have a DataID you'd need to create one. Actually that might not be that bad if you used the default id config (`satpy.dataset.dataid.default_id_keys_config`).

:thinking: Ok that might actually be cleaner. If in the `DatasetDict.__setitem__` you create a dictionary of all the metadata, then if there was a DataID in that metadata, pop it. If DataID get the `.id_keys` else use `default_id_keys_config`. Then create a new `DataID(id_keys, **attrs)` and use that. Maybe?